### PR TITLE
feat(kpagination/ktable): support offset-based pagination

### PR DESCRIPTION
**Pull request by dustinryerson**
# Summary
Adds support for offset-based pagination.

In practice, the following would need to be done in order to enable offset-based pagination:
1. Use the new `paginationType` prop for KTable: `<KTable :pagination-type="offset" />`
2. Use the new `offset` param in the fetcher function passed to KTable: `fetcher ({ offset }) {}`

**Assumptions:**
This implementation assumes that the API endpoint used in the fetcher function returns the next offset value in the response. Here is an example of the assumed format:
```json
{
  "data": [{}],
  "pagination": {
    "offset": "cGFnZS1udW1iZXI6Mg"
  },
}
```

This implementation also assumes that the API endpoint used in the fetcher function properly handles offset usage. Here's an example:
`GET /api/runtime_groups/6e30594e-9a90-4e0f-94f8-90b923696c83/services?size=100&offset=cGFnZS1udW1iZXI6Mg`

## Screenshot:
Screen Shot 2022-06-22 at 9 24 15 AM

## Vue 3

**Did you create a corresponding PR to add this change to the `beta` branch? (required)**

- [ ] **Yes**, here is a link to the PR: `[link to beta branch PR]`
- [x] **No**, I did not create a PR for the `beta` branch and I have added the `port to beta branch` tag to this PR.
- [ ] **No**, because `[type your reasons]`

If you have questions, tag `@adamdehaven` or `@kaiarrowood`.

---

## PR Checklist

- [x] Does not introduce dependencies
- [x] **Functional:** all changes do not break existing APIs and if so, bump major version.
- [x] **Tests pass:** check the output of yarn test packages/<Kongponent>
- [x] **Naming:** the files and the method and prop variables use the same naming conventions as other Kongponents
- [x] **Framework style:** abides by the essential rules in Vue's style guide
- [x] **Cleanliness:** does not have formatting issues, unused code (e.g., console.logs), or leftover comments
- [x] **Docs:** includes a technically accurate README, uses JSDOC where appropriate

### CODE REVIEW

File: packages/KPagination/KPagination.vue [hunk 1]
```diff
@@ -157,6 +168,18 @@ export default {
     testMode: {
       type: Boolean,
       default: false
+    },
+    paginationType: {
+      type: String,
+      default: 'default'
+    },
```
Comment (adamdehaven) [inline packages/KPagination/KPagination.vue:175]: We should add a `PropType` interface here for the `beta` branch port

File: packages/KTable/KTable.vue [hunk 2]
```diff
@@ -463,6 +468,10 @@ export default defineComponent({
     sortHandlerFn: {
       type: Function,
       default: () => ({})
+    },
+    paginationType: {
+      type: String,
+      default: 'default'
```
Comment (adamdehaven) [inline packages/KTable/KTable.vue:474]: We should add a type interface here for the `beta` branch port

File: packages/KPagination/KPagination.vue (hunk 1 above)
Comment (dustinryerson) [reply packages/KPagination/KPagination.vue:175]: I've added this to my beta branch port 👍

File: packages/KTable/KTable.vue (hunk 2 above)
Comment (dustinryerson) [reply packages/KTable/KTable.vue:474]: I've added this to my beta branch port 👍

**Comment by dustinryerson** [comment]
@kaiarrowood @adamdehaven I've created a PR for the `beta` branch: https://github.com/Kong/kongponents/pull/665

I'll update it with any changes requested for this PR.

File: docs/components/table.md [hunk 3]
```diff
@@ -243,6 +243,7 @@ object supporting the following properties:
 - Pagination support:
   - `page`: the currently visible page - starts at `1`
   - `pageSize`: the number of items to display per page
+  - `offset`: the value of the offset for offset-based pagination. `offset` **MUST** be included in the fetcher params for offset-based pagination to work properly.
```
Comment (kaiarrowood) [inline docs/components/table.md:246]: These doc updates need to also be copied over to KCatalog

File: docs/components/table.md [hunk 4]
```diff
@@ -483,6 +484,10 @@ Pass in an array of page sizes for the page size dropdown. If not provided will
 [15, 30, 50, 75, 100]
 ```
 
+### paginationType
+
+Pass in the type of pagination to be used. Options are `default` (page/pageSize) or `offset` (offset/pageSize)
```
Comment (kaiarrowood) [inline docs/components/table.md:489]: Can we show an example of what this looks like in the docs?

**Review by kaiarrowood — CHANGES_REQUESTED** [review]
Something is wrong with regular pagination alignment:

image

We should update this comment with an example of getting the `offset` from the `payload`

image

File: packages/KPagination/PaginationOffset.vue [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      default: false
+    },
+    nextButtonDisabled: {
+      type: Boolean,
+      default: false
+    },
+    /**
+     * Test mode - for testing only, strips out generated ids
+     */
+    testMode: {
```
[the remark this answers is not in the document]
Comment (adamdehaven) [reply packages/KPagination/PaginationOffset.vue:60]: If you remove, don't forget the tests

File: packages/KTable/KTable.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (adamdehaven) [reply packages/KTable/KTable.vue:474]: This should also be typed (the `paginationType` prop) and then you can use the type in the validator as well

File: docs/components/table.md (hunk 4 above)
Comment (dustinryerson) [reply docs/components/table.md:490]: Done 👍

File: packages/KPagination/KPagination.vue [hunk 6]
```diff
@@ -333,98 +356,97 @@ export default {
 }
 </script>
 
-<style lang="scss" scoped>
+<style lang="scss">
```
[the remark this answers is not in the document]
Comment (dustinryerson) [reply packages/KPagination/KPagination.vue]: Fixed 👍

File: docs/components/table.md (hunk 3 above)
Comment (dustinryerson) [reply docs/components/table.md:247]: We don't currently have a use case for KCatalog, so I haven't added support for using offset-based pagination to KCatalog yet.

**Comment by dustinryerson** [comment]
@kaiarrowood @adamdehaven thank you for the initial review! I've pushed up fixes and pr feedback changes, so this is ready for another round of reviews.

**Review by kaiarrowood — APPROVED** [review]
LGTM! 🐉 🔥

**Pull request merged**